<launch>
   
   <!--
      Demo of outdoor stereo mapping. 
      From bag: 
      $ rosbag record 
            /stereo_camera/left/image_raw_throttle/compressed 
            /stereo_camera/right/image_raw_throttle/compressed 
            /stereo_camera/left/camera_info_throttle 
            /stereo_camera/right/camera_info_throttle 
            /tf
    
      $ roslaunch rtabmap demo_stereo_outdoor.launch
      $ rosbag play -.-clock stereo_oudoorA.bag
   -->
   
   <!-- Choose visualization -->
   <arg name="rviz" default="true" />
   <arg name="rtabmapviz" default="false" />
   <arg name="local_bundle" default="true" />
   <arg name="stereo_sync" default="false" />
    
   <param name="use_sim_time" type="bool" value="True"/>
   
   <!-- Just to uncompress images for stereo_image_rect -->
   <node name="republish_left"  type="republish" pkg="image_transport" args="compressed in:=/stereo_camera/left/image_raw_throttle raw out:=/stereo_camera/left/image_raw_throttle_relay" />
   <node name="republish_right" type="republish" pkg="image_transport" args="compressed in:=/stereo_camera/right/image_raw_throttle raw out:=/stereo_camera/right/image_raw_throttle_relay" />

   <!-- Run the ROS package stereo_image_proc for image rectification -->
   <group ns="/stereo_camera" >
      <node pkg="stereo_image_proc" type="stereo_image_proc" name="stereo_image_proc">
         <remap from="left/image_raw"    to="left/image_raw_throttle_relay"/>
         <remap from="left/camera_info"  to="left/camera_info_throttle"/>
         <remap from="right/image_raw"   to="right/image_raw_throttle_relay"/>
         <remap from="right/camera_info" to="right/camera_info_throttle"/>
         <param name="disparity_range" value="128"/>
      </node>
      
      <node if="$(arg stereo_sync)" pkg="nodelet" type="nodelet" name="stereo_sync" args="standalone rtabmap_ros/stereo_sync">
      <remap from="left/image_rect"   to="left/image_rect_color"/>
      <remap from="right/image_rect"   to="right/image_rect"/>
      <remap from="left/camera_info"   to="left/camera_info_throttle"/>
      <remap from="right/camera_info"   to="right/camera_info_throttle"/>
    </node>
   </group>
              
   <group ns="rtabmap">   
   
      <!-- Stereo Odometry -->   
      <node pkg="rtabmap_ros" type="stereo_odometry" name="stereo_odometry" output="screen">
         <remap from="left/image_rect"       to="/stereo_camera/left/image_rect"/>
         <remap from="right/image_rect"      to="/stereo_camera/right/image_rect"/>
         <remap from="left/camera_info"      to="/stereo_camera/left/camera_info_throttle"/>
         <remap from="right/camera_info"     to="/stereo_camera/right/camera_info_throttle"/>
         <remap from="rgbd_image"            to="/stereo_camera/rgbd_image"/>
         <remap from="odom"                  to="/stereo_odometry"/>

         <param name="subscribe_rgbd"  type="bool" value="$(arg stereo_sync)"/>
         <param name="frame_id"        type="string" value="base_footprint"/>
         <param name="odom_frame_id"   type="string" value="odom"/>

         <param name="Odom/Strategy"      type="string" value="0"/> <!-- 0=Frame-to-Map, 1=Frame=to=Frame -->
         <param name="Vis/EstimationType" type="string" value="1"/> <!-- 0=3D->3D 1=3D->2D (PnP) -->
         <param name="Vis/MaxDepth"       type="string" value="0"/>
         <param name="Odom/GuessMotion" type="string" value="true"/>
         <param name="Vis/MinInliers"     type="string" value="10"/>
         <param unless="$(arg local_bundle)" name="OdomF2M/BundleAdjustment" type="string" value="0"/>
         <param name="OdomF2M/MaxSize"    type="string" value="1000"/> 
         <param name="GFTT/MinDistance"   type="string" value="10"/>
         <param name="GFTT/QualityLevel"  type="string" value="0.00001"/> 
         <param name="GFTT/QualityLevel"  type="string" value="0.00001"/>
      </node>
   
      <!-- Visual SLAM: args: "delete_db_on_start" and "udebug" -->
      <node name="rtabmap" pkg="rtabmap_ros" type="rtabmap" output="screen" args="--delete_db_on_start">
         <param name="frame_id"         type="string" value="base_footprint"/>
         <param unless="$(arg stereo_sync)" name="subscribe_stereo" type="bool" value="true"/>
         <param name="subscribe_depth"  type="bool" value="false"/>
         <param name="subscribe_rgbd"      type="bool" value="$(arg stereo_sync)"/>

         <remap from="left/image_rect"   to="/stereo_camera/left/image_rect_color"/>
         <remap from="right/image_rect"  to="/stereo_camera/right/image_rect"/>
         <remap from="left/camera_info"  to="/stereo_camera/left/camera_info_throttle"/>
         <remap from="right/camera_info" to="/stereo_camera/right/camera_info_throttle"/>
         <remap from="rgbd_image"        to="/stereo_camera/rgbd_image"/>

         <remap from="odom" to="/stereo_odometry"/>

         <param name="queue_size" type="int" value="30"/>
         <param name="map_negative_poses_ignored" type="bool" value="true"/>

         <!-- RTAB-Map's parameters -->
         <param name="Rtabmap/TimeThr"                   type="string" value="700"/>
         <param name="Grid/DepthDecimation"              type="string" value="4"/>
         <param name="Grid/FlatObstacleDetected"         type="string" value="true"/>
         <param name="Kp/MaxDepth"                       type="string" value="0"/>
         <param name="Kp/DetectorStrategy"               type="string" value="6"/>  
         <param name="Vis/EstimationType"                type="string" value="1"/>   <!-- 0=3D->3D, 1=3D->2D (PnP) -->
         <param name="Vis/MaxDepth"                      type="string" value="0"/>
         <param name="RGBD/CreateOccupancyGrid"          type="string" value="true"/>       
      </node>
      
      <!-- Visualisation RTAB-Map -->
      <node if="$(arg rtabmapviz)" pkg="rtabmap_ros" type="rtabmapviz" name="rtabmapviz" args="-d $(find rtabmap_ros)/launch/config/rgbd_gui.ini" output="screen">
         <param unless="$(arg stereo_sync)" name="subscribe_stereo"    type="bool" value="true"/>
         <param name="subscribe_odom_info" type="bool" value="true"/>
         <param name="subscribe_rgbd"      type="bool" value="$(arg stereo_sync)"/>
         <param name="queue_size"          type="int" value="10"/>
         <param name="frame_id"            type="string" value="base_footprint"/>
         
         <remap from="left/image_rect"  to="/stereo_camera/left/image_rect_color"/>
         <remap from="right/image_rect"  to="/stereo_camera/right/image_rect"/>
         <remap from="left/camera_info"  to="/stereo_camera/left/camera_info_throttle"/>
         <remap from="right/camera_info" to="/stereo_camera/right/camera_info_throttle"/>
         <remap from="rgbd_image"          to="/stereo_camera/rgbd_image"/>
         <remap from="odom_info"         to="odom_info"/>
         <remap from="odom"              to="/stereo_odometry"/>
         <remap from="mapData"           to="mapData"/>
      </node>
         
   </group>
  
   <!-- Visualisation RVIZ --> 
   <node if="$(arg rviz)" pkg="rviz" type="rviz" name="rviz" args="-d $(find rtabmap_ros)/launch/config/demo_stereo_outdoor.rviz"/>

</launch>
